dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1022 @@
+
+#if NETCOREAPP3_0_OR_GREATER
+
+    public async Task<WorkflowHistory> FetchHistoryAsync(
+        WorkflowHistoryEventFilterType eventFilterType = WorkflowHistoryEventFilterType.AllEvents,
+        bool skipArchival = false,
+        RPCOptions? options = null) { /*...*/ }
+
+    public IAsyncEnumerator<HistoryEvent> FetchHistoryEvents(
+        bool waitNewEvent = false,
+        WorkflowHistoryEventFilterType eventFilterType = WorkflowHistoryEventFilterType.AllEvents,
+        bool skipArchival = false,
+        RPCOptions? options = null) { /*...*/ }
+
+#endif
+}
+```
+
+Notes:
+
+* Notice you can provide a return type explicitly on the `GetResultAsync` call if the generic one on the handle isn't
+  what you want
+* Like start workflow, signal/query string-based calls accept an array of objects instead of trying to use varargs or
+  separating single arg vs multi. This makes the overload count better and will give users compile time errors if it's
+  not literally an object array
+* Notice that `TerminateAsync` does not _require_ `args` like start, signal, and query calls that accept `args` arrays
+  because we don't think it's important for users to explicitly send empty if they don't want them like we think they
+  should for the others
+* Same issue with `FetchHistory*` as `ListWorkflows` on client - no `IAsyncEnumerator` on really old .NET versions, but
+  we're ok with that
+
+## Worker
+
+Usage:
+
+```csharp
+using Temporal.Client;
+using Temporal.Worker;
+using MyNamespace;
+
+// Connect
+var client = await TemporalClient.ConnectAsync(new("localhost:7233"));
+
+// Build worker options
+var options = new TemporalWorkerOptions("task queue");
+options.AddActivities(new GreetingActivities());
+
+// Run until Ctrl+C
+var worker = new TemporalWorker(client, options);
+var cts = new CancellationTokenSource();
+Console.CancelKeyPress += (sender, eventArgs) => {
+    eventArgs.Cancel = true;
+    cts.Cancel();
+};
+await worker.ExecuteAsync(ctx.Token);
+```
+
+Looks like:
+
+```csharp
+namespace Temporal.Worker;
+
+using Temporal.Activity;
+
+public class TemporalWorkerOptions
+{
+    public TemporalWorkerOptions() { /*...*/ }
+    public TemporalWorkerOptions(string taskQueue) { /*...*/ }
+
+    public string? TaskQueue { get; set; }
+
+    public List<ActivityDefinition> ActivityDefinitions { get; set; } = new();
+
+    public void AddActivities(params object[] activities) { /*...*/ }
+
+    public void AddActivityDefinition(ActivityDefinition definition) { /*...*/ }
+
+    // Many more options not listed
+}
+
+public class TemporalWorker
+{
+    public TemporalWorker(IWorkerClient client, TemporalWorkerOptions options) { /*...*/ }
+
+    public Task ExecuteAsync(CancellationToken stoppingToken) { /*...*/ }
+}
+```
+
+Notes:
+
+* Like other options, technically all worker options are optional at compile time but we fail at runtime if task queue
+  is not set
```

Comment:
If this is the case why have the no-arg constructor at all?

Reply:
Mentioned way later in document, but this follows https://learn.microsoft.com/en-us/dotnet/core/extensions/options which states:

>  When using the options pattern, an options class:
> * Must be non-abstract with a public parameterless constructor

I did bring this up in the `#dotnet-sdk` channel on public Slack on whether to choose this pattern or immutable records (my personal favorite), but this is the safest bet for all of our options.